### ISSUE TITLE
The real time chatting feature of the program

### ISSUE THREAD

**Issue by yahuiliang** [superyang713/ETL#3]
The real-time chatting is a mandatory feature of the program. However, it is not determined yet how to implement that. If you have any thoughts, observations, or experiences about how to implement this feature, feel free to leave a comment below. You can put the information in any formats: plain text, links, pictures, etc. All comments will be considered. Thanks for your help!

**Comment by superyang713** [comment]
you can take a look at this repo https://github.com/dondido/webrtc-video-room. I think the best way to implement this is to use webrtc protocol so we can eliminate the need of the third party server.

**Comment by yahuiliang** [comment]
@superyang713 Thanks for your advising. I took a look at webRTC yesterday, and I feel this is the good way to set up the P2P video streaming between users. I am trying to set up the demo on the website these days. Hopefully I can get this done by Friday midnight.

**Comment by superyang713** [comment]
yes. Get the prototype working and we will implement it in our app. Thank you.

**Comment by yahuiliang** [comment]
I tried to implement the webRTC demo into the application based on the [instruction](https://codelabs.developers.google.com/codelabs/webrtc-web/#7)  and I got the basic video peer-to-peer video streaming working. Here is the [link](https://github.com/yahuiliang/ETL/tree/video-streaming) to my branch. However, it just worked when you open the page for the first time. I think it will be an easy fix. In order to run what I have there, use `node server` to start the socket server first. Then `npm start` to start the front end server.

Here are what I learned and what I still feel confused about these days:

WebRTC consists of 4 servers. 
1. TURN (for relay purpose) 
    1.1 I do not actually know what it actually does in the application
2. STUN (for retrieving the peer ip)
3. React server (for handling frontend requests)
4. Signaling server (for exchanging meta data between peers, and this is what I implemented in server.js)

I still do not completely understand how these servers cooperate with each other. Especially between TURN and STUN. I think my code still use the local stuff to exchange the information.

**Comment by yahuiliang** [comment]
@superyang713 Got the demo working now. If you pull my branch, do `node server.js` first in a terminal to start the signaling server, and then `npm start` in another terminal to start up the React server. Then create two users which are student and teacher. Open the website on two pages, and login using different roles. Open the video tab and try to stream videos to your self. Make sure to use `localhost` to test it since the camera can only be opened in either `https` connection and `localhost` connection. I have not added audio to it since it brought some confusions on testing. However, I am pretty sure that it is easy to add the audio track to the connection
